src/app/browser/qgsinbuiltdataitemproviders.cpp

```diff
@@ -366,6 +366,25 @@ void QgsLayerItemGuiProvider::populateContextMenu( QgsDataItem *item, QMenu *men
   } );
   menu->addAction( addAction );
 
+  if ( item->capabilities2() & QgsDataItem::Delete )
+  {
+    QList<QgsLayerItem *> selectedDeletableItems;
+    for ( QgsDataItem *selectedItem : selectedItems )
+    {
+      if ( qobject_cast<QgsLayerItem *>( selectedItem ) && ( selectedItem->capabilities2() & QgsDataItem::Delete ) )
+        selectedDeletableItems.append( qobject_cast<QgsLayerItem *>( selectedItem ) );
+    }
+
+    const QString deleteText = selectedDeletableItems.count() == 1 ? tr( "Delete Layer" )
+                               : tr( "Delete Selected Layers" );
+    QAction *deleteAction = new QAction( deleteText, menu );
+    connect( deleteAction, &QAction::triggered, this, [ = ]
+    {
+      deleteLayers( selectedDeletableItems );
```

Comment:
This is likely to result in crashes, because after the first table is deleted providers may (and usually will) trigger a refresh of their parent. This happens in a background thread, and may result in items within the selectedDeletableItems list being deleted before deleteLayers has had a chance to handle them. I think what should be done here is that selectedDeletableItems will need to be a string list, storing the paths of the selected items instead of the items themselves. (I've experienced this crash myself when deleting OGR items, which was why the OGR provider was adapted to use a static deleteLayer method with strings instead of the layer item).

Reply:
And my apologies, I should have picked this up in the earlier review. That's my fault.

Reply:
Hm... Good point.
Can't we just check, if the item is still valid before we call to delete it?
If I understand you correctly, it means, that it shouldn't be implemented in the specific layer items and have a general `deleteItemAccordingUri( item-uri, item-type )` function in `qgsinbuiltdataitemproviders`?

Reply:
> This is likely to result in crashes, because after the first table is deleted providers may (and usually will) trigger a refresh of their parent. This happens in a background thread, and may result in items within the selectedDeletableItems list being deleted before deleteLayers has had a chance to handle them.

I'm not sure I can follow this logic @nyalldawson 
 
Does this mean, a background thread deletes items from the UI? If yes, this sounds like a guarantee for crashes and should be fixed directly instead of working around potential issues here.
Instead they should be deleted (via queued connection) on the main thread, in which case they will be deleted after this code has been executed (because this code is triggered by the UI on the main thread and executed via direct connection unless I'm mistaken).

Reply:
> Instead they should be deleted (via queued connection) on the main thread

I believe this is already the case

> in which case they will be deleted after this code has been executed (because this code is triggered by the UI on the main thread and executed via direct connection unless I'm mistaken).

Ideally yes, but in practice there must be some processEvents calls or something messing things up, because I *definitely* got easily reproducible crashes when this approach was used in the OGR provider.

Reply:
Is there a stacktrace around for this or instructions how to reproduce this?

To be honest, I'm not a big fan of implementing workarounds for badly designed code without seeing it first and having a good ground to decide that fixing the broken code is (much) harder than implementing the workaround. :/

Reply:
It's https://github.com/qgis/QGIS/pull/8165/commits/7ad1d29a7dd083e827a16d48d06c501d875a6582 -- and that commit gives more clues, because it's while the confirmation dialog is shown (main thread) that the main thread event loop is running and deleting the items due to the previously queued refresh. 

So that's all good -- no nasty bugs in master which we're hiding. But this PR also has messageboxes showing while deleting layers (confirmation for each layer, failure warning) and so needs to handle that same case. Approaches would be encapsulating the layer parameters from all items in advance and only using these during the delete, OR

1. show a single confirmation once in advance showing a summary of all layers to be deleted, and not show a confirmation for each individual layer in turn, and 
2. only show the messagebox failure message after all items have been processed